### ISSUE TITLE
File.Copy no longer respects FileShare.None on Linux/Mac in 5.0

### ISSUE THREAD

**Issue by nguerrera** [dotnet/runtime#31429]
Found this taking a new runtime into dotnet/sdk test run environment, as one of our tests started failing on Linux/Mac.

### Program.cs
``` C#
using System;
using System.IO;

class Program
{
    static void Main()
    { 
        File.WriteAllText("x", "x");
        File.WriteAllText("y", "y");

        using (var stream = new FileStream("y", FileMode.Open, FileAccess.Read, FileShare.None))
        {
            File.Copy("x", "y", overwrite: true);
        }
        
        Console.WriteLine("Success");
    }
}
```

### Repro.csproj
``` xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>netcoreapp3.0;netcoreapp5.0</TargetFrameworks>
  </PropertyGroup>
</Project>
```

### Repro steps
* dotnet run -f netcoreapp3.0
* dotnet run -f netcoreapp5.0

### Expected result
Both 3.0 and 5.0 fail with IOExceptions

### Actual result
Only 3.0 fails, 5.0 succeeds unexpectedly

#### 3.0 (fails as expected)

```/
Unhandled exception. System.IO.IOException: The process cannot access the file '/home/nicholg/src/sdk/q/y' because it is being used by another process.
   at System.IO.FileStream.Init(FileMode mode, FileShare share, String originalPath)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)      
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at Program.Main()
```

#### 5.0 (succeeds unexpectedly)
```
Success
```

@stephentoub @JeremyKuhne

**Comment by JeremyKuhne** [comment]
@filipnavara Any luck on a fix?

**Comment by filipnavara** [comment]
@JeremyKuhne I submitted a PR the same day as my last comment (https://github.com/dotnet/corefx/pull/42458) but it was not merged because it was making duplicated code (on native and managed side) which neither I nor @stephentoub liked. Currently there's PR #1495 which should also fix it in slightly cleaner way. Unfortunately it still requires some tweaking to pass the tests and there was no feedback on its last version.

**Comment by danmosemsft** [comment]
From @stephentoub : We broke this in 5.0, we should fix it in 5.0, either reverting the change that broke it (https://github.com/dotnet/corefx/pull/37583) or finishing the PR that fixes it (https://github.com/dotnet/runtime/pull/33159).

**Comment by carlossanlop** [comment]
@Jozkee will help revert the change
